File: eldoria/data/home_region_content.py
"""The 15 hand-authored, tightly connected home-region locations (Oakhaven Village and surroundings).

Spliced onto the deterministically-generated start city (see world_generator's
`graft` call) so every new game keeps this same tested content while the much
larger procedural world remains reachable beyond it -- Oakhaven's `west` exit
is the one deliberate gateway out.

Deliberately NOT wired through the procedural SideQuestKind machinery: these
quests are fetch-quests gated on specific items or kill counts, resolved
through a yes/no accept during `talk` and an automatic turn-in on a later
`talk` -- driven from commands.py's `talk()` using plain string quest ids in
GameSession.active_home_region_quests/completed_side_quests.
"""
from __future__ import annotations

import random
from dataclasses import replace

from eldoria.models import (
    Biome,
    Disposition,
    GameLocation,
    Item,
    ItemKind,
    PopulationTier,
    SpawnEntry,
    SpawnKind,
    StatusEffect,
    TerrainKind,
)
from eldoria.data.ai_companion_lore import MR_DAVIS_NAME
from eldoria.world import stat_generator as sg

# --- Quest ids (GameSession.active_home_region_quests / completed_side_quests keys) ---
QUEST_ANCIENT_RELIC = "ancient_relic"
QUEST_POISONED_WATERS = "poisoned_waters"
QUEST_GOBLIN_OUTBREAK = "goblin_outbreak"
QUEST_LOST_CARGO = "lost_cargo"
QUEST_MOUNTAIN_RESCUE = "mountain_rescue"
QUEST_SLAY_THE_WYRM = "slay_the_wyrm"
QUEST_ALCHEMICAL_FUNGI = "alchemical_fungi"
QUEST_ANCIENT_COMPASS = "ancient_compass"

QUEST_TITLES: dict[str, str] = {
    QUEST_ANCIENT_RELIC: "The Ancient Relic",
    QUEST_POISONED_WATERS: "The Poisoned Waters",
    QUEST_GOBLIN_OUTBREAK: "Goblin Outbreak",
    QUEST_LOST_CARGO: "Lost Cargo",
    QUEST_MOUNTAIN_RESCUE: "Mountain Rescue",
    QUEST_SLAY_THE_WYRM: "Slay the Wyrm",
    QUEST_ALCHEMICAL_FUNGI: "Alchemical Fungi",
    QUEST_ANCIENT_COMPASS: "The Ancient Compass",
}

# --- Named NPCs (exact names commands.py's handle_home_region_npc matches on) ---
ELDER_THERON = "Elder Theron"
FISHERMAN_FINN = "Fisherman Finn"
MOUNTAIN_GUIDE = "Mountain Guide"
ARCANE_VENDOR = "Arcane Vendor"
ANCIENT_SCHOLAR = "Ancient Scholar"
LOST_MINER = "Lost Miner"

# Trader names deliberately contain a dialogue_content TRADER keyword
# (trader/merchant/keeper/vendor) so commands.py's open_shop() finds them
# with zero changes -- see world/shop_generator.py's doc.
OAKHAVEN_MERCHANT = "Oakhaven General Merchant"
COASTAL_TRADER = "Coastal Market Trader"
MOUNTAIN_KEEPER = "Ironstone Outpost Keeper"

# --- Quest item names (checked by name in commands.py's handle_home_region_npc) ---
ITEM_ANCIENT_RELIC = "Ancient Relic"
ITEM_SWAMP_HERB = "Swamp Herb"
ITEM_SHIP_MANIFEST = "Ship's Manifest"
ITEM_LOST_MINER_NOTE = "Lost Miner's Note"
ITEM_ANCIENT_COMPASS = "Ancient Compass"
ITEM_GLOWING_MUSHROOM = "Glowing Mushroom"
SWAMP_HERB_REQUIRED = 3
GLOWING_MUSHROOM_REQUIRED = 2
GOBLINS_REQUIRED = 3


def _enemy(name: str, tier: int, seed: int, resistances: frozenset[StatusEffect] = frozenset()) -> SpawnEntry:
    stats = sg.creature_stats(tier, random.Random(seed))
    if resistances:
        stats = replace(stats, status_resistances=frozenset(resistances))
    return SpawnEntry(name=name, kind=SpawnKind.CREATURE, disposition=Disposition.HOSTILE, stats=stats)


def _npc(name: str, seed: int, tier: int = 1) -> SpawnEntry:
    return SpawnEntry(
        name=name, kind=SpawnKind.NPC, disposition=Disposition.PASSIVE,
        stats=sg.creature_stats(tier, random.Random(seed)),
    )


def _goblin(seed: int) -> SpawnEntry:
    return _enemy("Goblin Scavenger", 1, seed)


def _dire_wolf(seed: int) -> SpawnEntry:
    return _enemy("Dire Wolf", 1, seed)


def _mountain_goat(seed: int) -> SpawnEntry:
    return _enemy("Mountain Goat", 1, seed)


def _giant_toad(seed: int) -> SpawnEntry:
    return _enemy("Giant Toad", 2, seed, frozenset({StatusEffect.BURN}))


def _bandit_outlaw(seed: int) -> SpawnEntry:
    return _enemy("Bandit Outlaw", 2, seed)


def _giant_crab(seed: int) -> SpawnEntry:
    return _enemy("Giant Crab", 2, seed, frozenset({StatusEffect.POISON}))


def _shadow_cultist(seed: int) -> SpawnEntry:
    return _enemy("Shadow Cultist", 3, seed)


def _skeletal_warrior(seed: int) -> SpawnEntry:
    return _enemy("Skeletal Warrior", 3, seed, frozenset({StatusEffect.POISON}))


def _stone_golem(seed: int) -> SpawnEntry:
    return _enemy("Stone Golem", 4, seed, frozenset({StatusEffect.POISON, StatusEffect.BURN}))


def _ancient_dragon(seed: int) -> SpawnEntry:
    return _enemy("Ancient Flame Dragon", 5, seed, frozenset({StatusEffect.BURN}))


def _quest_item(name: str, tier: int, seed: int) -> Item:
    return sg.quest_item(name, tier, random.Random(seed))


def _material(name: str, seed: int) -> Item:
    item = sg.quest_item(name, 1, random.Random(seed))
    return replace(item, kind=ItemKind.MATERIAL, value=5)


def graft(locations: dict[str, GameLocation], anchor_x: int, anchor_y: int) -> None:
    """Splices the 15 home-region locations into an already-generated location map, anchored at (anchor_x, anchor_y).

    Offsets are unique but NOT geometry-consistent with each location's
    cardinal exits (the map view is cosmetic-only, non-load-bearing); every
    location's exits map is hand-set to the original topology regardless of
    where it actually sits on the grid.

    Exits intentionally point INTO 14 ids this function itself creates -- a
    fully closed loop -- except Oakhaven's `west`, which is deliberately left
    pointing at whatever real generated tile already existed there before the
    graft: the one gateway from the tested home region out into the much
    larger procedural world.
    """

    def loc_id(dx: int, dy: int) -> str:
        return f"{anchor_x + dx}_{anchor_y + dy}"

    id_oakhaven = loc_id(0, 0)
    id_whispering_woods = loc_id(0, -1)
    id_deep_woods = loc_id(0, -2)
    id_shadow_caves = loc_id(1, -1)
    id_sunken_citadel = loc_id(2, -1)
    id_old_road = loc_id(1, 0)
    id_ironstone_foothills = loc_id(2, 0)
    id_ironstone_pass = loc_id(2, 1)
    id_dragons_peak = loc_id(2, 2)
    id_eastern_swamps = loc_id(0, 1)
    id_coastal_town = loc_id(1, 1)
    id_sunken_shipwreck = loc_id(1, 2)
    id_mountain_pass_north = loc_id(3, 0)
    id_ancient_ruins = loc_id(-1, -2)
    id_forgotten_crypt = loc_id(-1, -3)

    def place(
        loc_id_: str, dx: int, dy: int, name: str, description: str,
        exits: dict[str, str], population_tier: PopulationTier, difficulty_tier: int,
        beings: tuple[SpawnEntry, ...] = (), items: tuple[Item, ...] = (),
    ) -> None:
        locations[loc_id_] = GameLocation(
            id=loc_id_, x=anchor_x + dx, y=anchor_y + dy, biome=Biome.PLAINS,
            name=name, description=description, population_tier=population_tier,
            difficulty_tier=difficulty_tier, difficulty_score=difficulty_tier * 20,
            beings=beings, exits=exits, terrain=TerrainKind.LAND, items=items,
        )

    existing = locations.get(loc_id(-1, 0))
    existing_west_neighbor = existing.id if existing is not None else id_oakhaven

    place(
        id_oakhaven, 0, 0, "Oakhaven Village",
        "A serene village surrounding a central square. The air smells of fresh bread and damp earth.",
        {"north": id_whispering_woods, "east": id_old_road, "south": id_eastern_swamps, "west": existing_west_neighbor},
        PopulationTier.CITY, 1,
        beings=(_npc(ELDER_THERON, seed=101), _npc(OAKHAVEN_MERCHANT, seed=102), _npc(MR_DAVIS_NAME, seed=110)),
    )
    place(
        id_whispering_woods, 0, -1, "Whispering Woods",
        "Dense forest with howling wind through misty trees. Ancient, gnarled trees loom overhead.",
        {"south": id_oakhaven, "north": id_deep_woods, "east": id_shadow_caves},
        PopulationTier.WILDERNESS, 1,
        beings=(_dire_wolf(201), _goblin(202)),
    )
    place(
        id_deep_woods, 0, -2, "Deep Whispering Woods",
        "Pitch dark canopy where sunlight cannot reach. Strange sounds echo from the shadows.",
        {"south": id_whispering_woods, "west": id_ancient_ruins},
        PopulationTier.WILDERNESS, 2,
        beings=(_goblin(203), _dire_wolf(204)),
        items=(_quest_item(ITEM_ANCIENT_RELIC, 2, 301),),
    )
    place(
        id_shadow_caves, 1, -1, "Shadow Caves",
        "Damp limestone cave network dripping with glowing flora. The air is cool and still.",
        {"west": id_whispering_woods, "east": id_sunken_citadel},
        PopulationTier.WILDERNESS, 2,
        beings=(_goblin(205), _giant_toad(206)),
        items=(_material(ITEM_GLOWING_MUSHROOM, 302), _material(ITEM_GLOWING_MUSHROOM, 303)),
    )
    place(
        id_sunken_citadel, 2, -1, "Sunken Citadel Courtyard",
        "Ancient flooded stone fortress radiating dark power. Water laps at crumbling walls.",
        {"west": id_shadow_caves},
        PopulationTier.CITY, 3,
        beings=(_npc(ARCANE_VENDOR, seed=103), _shadow_cultist(207), _skeletal_warrior(208)),
    )
    place(
        id_old_road, 1, 0, "Old Forest Road",
        "Long dirt path connecting distant provinces. Wagon tracks are faintly visible.",
        {"west": id_oakhaven, "east": id_ironstone_foothills, "south": id_coastal_town},
        PopulationTier.WILDERNESS, 1,
        beings=(_goblin(209), _bandit_outlaw(210)),
    )
    place(
        id_ironstone_foothills, 2, 0, "Ironstone Foothills",
        "Rocky ascending terrain surrounded by jagged cliff faces. The wind picks up here.",
        {"west": id_old_road, "north": id_ironstone_pass, "east": id_mountain_pass_north},
        PopulationTier.CITY, 2,
        beings=(_npc(MOUNTAIN_GUIDE, seed=104), _npc(MOUNTAIN_KEEPER, seed=105), _bandit_outlaw(211), _mountain_goat(212)),
    )
    place(
        id_ironstone_pass, 2, 1, "Ironstone Mountain Pass",
        "Freezing mountain path with howling blizzards. The air is thin and cold.",
        {"south": id_ironstone_foothills, "up": id_dragons_peak},
        PopulationTier.WILDERNESS, 4,
        beings=(_stone_golem(213), _mountain_goat(214)),
    )
    place(
        id_dragons_peak, 2, 2, "Dragon's Peak Summit",
        "High volcanic summit covered in ancient scorch marks. A faint smell of sulfur lingers.",
        {"down": id_ironstone_pass},
        PopulationTier.WILDERNESS, 5,
        beings=(_ancient_dragon(215),),
    )
    place(
        id_eastern_swamps, 0, 1, "Eastern Swamps",
        "A murky, humid marshland. Strange plants and buzzing insects fill the air.",
        {"north": id_oakhaven},
        PopulationTier.WILDERNESS, 1,
        beings=(_giant_toad(216), _dire_wolf(217)),
        items=tuple(_material(ITEM_SWAMP_HERB, 310 + i) for i in range(SWAMP_HERB_REQUIRED)),
    )
    place(
        id_coastal_town, 1, 1, "Coastal Town of Port Eldoria",
        "A bustling port town with the scent of salt and fish. Ships dock constantly.",
        {"north": id_old_road, "east": id_sunken_shipwreck},
        PopulationTier.CITY, 2,
        beings=(_npc(FISHERMAN_FINN, seed=106), _npc(COASTAL_TRADER, seed=107)),
    )
    place(
        id_sunken_shipwreck, 1, 2, "Sunken Shipwreck",
        "The remains of a grand ship, half-submerged in shallow waters.",
        {"west": id_coastal_town},
        PopulationTier.WILDERNESS, 2,
        beings=(_giant_crab(218),),
        items=(_quest_item(ITEM_SHIP_MANIFEST, 2, 304),),
    )
    place(
        id_mountain_pass_north, 3, 0, "Northern Mountain Pass",
        "A narrow, icy path winding through towering peaks.",
        {"west": id_ironstone_foothills, "north": id_forgotten_crypt},
        PopulationTier.WILDERNESS, 3,
        beings=(_mountain_goat(219), _stone_golem(220), _npc(LOST_MINER, seed=108)),
    )
    place(
        id_ancient_ruins, -1, -2, "Ancient Ruins",
        "Crumbling stone structures overgrown with vines.",
        {"east": id_deep_woods, "north": id_forgotten_crypt},
        PopulationTier.WILDERNESS, 3,
        beings=(_skeletal_warrior(221), _shadow_cultist(222), _npc(ANCIENT_SCHOLAR, seed=109)),
    )
    place(
        id_forgotten_crypt, -1, -3, "Forgotten Crypt",
        "A dark, musty crypt beneath the ancient ruins.",
        {"south": id_ancient_ruins, "west": id_mountain_pass_north},
        PopulationTier.WILDERNESS, 4,
        beings=(_skeletal_warrior(223),),
        items=(_quest_item(ITEM_ANCIENT_COMPASS, 3, 305),),
    )

File: core/src/commonMain/kotlin/eldoria/core/data/HomeRegionContent.kt
package eldoria.core.data

import eldoria.core.model.Biome
import eldoria.core.model.Disposition
import eldoria.core.model.GameLocation
import eldoria.core.model.Item
import eldoria.core.model.ItemKind
import eldoria.core.model.PopulationTier
import eldoria.core.model.SpawnEntry
import eldoria.core.model.SpawnKind
import eldoria.core.model.StatBlock
import eldoria.core.model.StatusEffect
import eldoria.core.model.TerrainKind
import eldoria.core.world.StatGenerator
import kotlin.random.Random

/**
 * The Python prototype's 15 hand-authored, tightly connected locations
 * (Oakhaven Village and its surrounding woods/mountains/coast), spliced
 * onto the deterministically-generated start city (see WorldGenerator's
 * `graft` call, right before it returns the World) so every new game keeps
 * this same tested content while Kotlin's much larger procedural world
 * remains reachable beyond it -- Oakhaven's `west` exit is the one
 * deliberate gateway out (see `graft`'s doc).
 *
 * Deliberately NOT wired through the procedural SideQuestKind/pickBeings
 * machinery: the Python source's quests are fetch-quests gated on specific
 * items or kill counts, resolved through a yes/no accept during `talk` and
 * an automatic turn-in on a later `talk` -- a different shape than
 * SideQuestKind's "resolve <keyword>" flavor-quest model (see
 * model/SideQuest.kt), and SpawnEntry.offersSideQuest is one-quest-per-NPC
 * while several of these NPCs (Elder Theron, Mountain Guide) give more
 * than one. Rather than force-fit either constraint, this is a small,
 * self-contained, faithfully-ported dialogue/quest system of its own,
 * driven from Game.kt's `talk()` (see `handleHomeRegionNpc`) using plain
 * string quest ids in GameSession.activeHomeRegionQuests/completedSideQuests
 * (see GameSession's doc on why NOT discoveredQuests).
 */
object HomeRegionContent {

    // --- Quest ids (GameSession.activeHomeRegionQuests / completedSideQuests keys) ---
    const val QUEST_ANCIENT_RELIC = "ancient_relic"
    const val QUEST_POISONED_WATERS = "poisoned_waters"
    const val QUEST_GOBLIN_OUTBREAK = "goblin_outbreak"
    const val QUEST_LOST_CARGO = "lost_cargo"
    const val QUEST_MOUNTAIN_RESCUE = "mountain_rescue"
    const val QUEST_SLAY_THE_WYRM = "slay_the_wyrm"
    const val QUEST_ALCHEMICAL_FUNGI = "alchemical_fungi"
    const val QUEST_ANCIENT_COMPASS = "ancient_compass"

    val questTitles: Map<String, String> = mapOf(
        QUEST_ANCIENT_RELIC to "The Ancient Relic",
        QUEST_POISONED_WATERS to "The Poisoned Waters",
        QUEST_GOBLIN_OUTBREAK to "Goblin Outbreak",
        QUEST_LOST_CARGO to "Lost Cargo",
        QUEST_MOUNTAIN_RESCUE to "Mountain Rescue",
        QUEST_SLAY_THE_WYRM to "Slay the Wyrm",
        QUEST_ALCHEMICAL_FUNGI to "Alchemical Fungi",
        QUEST_ANCIENT_COMPASS to "The Ancient Compass",
    )

    // --- Named NPCs (exact names Game.kt's handleHomeRegionNpc matches on) ---
    const val ELDER_THERON = "Elder Theron"
    const val FISHERMAN_FINN = "Fisherman Finn"
    const val MOUNTAIN_GUIDE = "Mountain Guide"
    const val ARCANE_VENDOR = "Arcane Vendor"
    const val ANCIENT_SCHOLAR = "Ancient Scholar"
    const val LOST_MINER = "Lost Miner"

    // Trader names deliberately contain a DialogueContentRegistry TRADER
    // keyword (trader/merchant/keeper/vendor) so Game.kt's openShop() finds
    // them with zero changes -- see world/ShopGenerator.kt's doc.
    const val OAKHAVEN_MERCHANT = "Oakhaven General Merchant"
    const val COASTAL_TRADER = "Coastal Market Trader"
    // Ironstone Outpost Keeper: the Python source defined a fourth shop
    // (`mountain_shop`) but never actually attached it to any location --
    // fixed here rather than ported faithfully, since leaving a fully
    // stocked shop permanently unreachable reads as an oversight, not
    // content worth preserving.
    const val MOUNTAIN_KEEPER = "Ironstone Outpost Keeper"

    // --- Quest item names (checked by name in Game.kt's handleHomeRegionNpc) ---
    const val ITEM_ANCIENT_RELIC = "Ancient Relic"
    const val ITEM_SWAMP_HERB = "Swamp Herb"
    const val ITEM_SHIP_MANIFEST = "Ship's Manifest"
    const val ITEM_LOST_MINER_NOTE = "Lost Miner's Note"
    const val ITEM_ANCIENT_COMPASS = "Ancient Compass"
    const val ITEM_GLOWING_MUSHROOM = "Glowing Mushroom"
    const val SWAMP_HERB_REQUIRED = 3
    const val GLOWING_MUSHROOM_REQUIRED = 2
    const val GOBLINS_REQUIRED = 3

    private fun enemy(name: String, tier: Int, seed: Long, resistances: Set<StatusEffect> = emptySet()): SpawnEntry =
        SpawnEntry(
            name = name,
            kind = SpawnKind.CREATURE,
            disposition = Disposition.HOSTILE,
            stats = StatGenerator.creatureStats(tier, Random(seed)).copy(statusResistances = resistances),
        )

    private fun npc(name: String, tier: Int = 1, seed: Long): SpawnEntry =
        SpawnEntry(
            name = name,
            kind = SpawnKind.NPC,
            disposition = Disposition.PASSIVE,
            stats = StatGenerator.creatureStats(tier, Random(seed)),
        )

    private fun goblin(seed: Long) = enemy("Goblin Scavenger", tier = 1, seed = seed)
    private fun direWolf(seed: Long) = enemy("Dire Wolf", tier = 1, seed = seed)
    private fun mountainGoat(seed: Long) = enemy("Mountain Goat", tier = 1, seed = seed)
    private fun giantToad(seed: Long) = enemy("Giant Toad", tier = 2, seed = seed, resistances = setOf(StatusEffect.BURN))
    private fun banditOutlaw(seed: Long) = enemy("Bandit Outlaw", tier = 2, seed = seed)
    private fun giantCrab(seed: Long) = enemy("Giant Crab", tier = 2, seed = seed, resistances = setOf(StatusEffect.POISON))
    private fun shadowCultist(seed: Long) = enemy("Shadow Cultist", tier = 3, seed = seed)
    private fun skeletalWarrior(seed: Long) = enemy("Skeletal Warrior", tier = 3, seed = seed, resistances = setOf(StatusEffect.POISON))
    private fun stoneGolem(seed: Long) = enemy("Stone Golem", tier = 4, seed = seed, resistances = setOf(StatusEffect.POISON, StatusEffect.BURN))
    private fun ancientDragon(seed: Long) = enemy("Ancient Flame Dragon", tier = 5, seed = seed, resistances = setOf(StatusEffect.BURN))

    private fun questItem(name: String, tier: Int, seed: Long): Item = StatGenerator.questItem(name, tier, Random(seed))
    private fun material(name: String, seed: Long): Item = StatGenerator.questItem(name, tier = 1, rng = Random(seed)).copy(kind = ItemKind.MATERIAL, value = 5)

    /**
     * Splices the 15 home-region locations into an already-generated
     * location map, anchored at (anchorX, anchorY) -- the same tile
     * Game.kt's start-city selection will land on (see WorldGenerator's
     * call site, which passes the deterministically-chosen start city's
     * own coordinates). Offsets are unique but NOT geometry-consistent
     * with each location's cardinal exits (MapRenderer's spatial display
     * is cosmetic-only, confirmed non-load-bearing -- see the Phase 2
     * research this was built from); every location's `exits` map is
     * hand-set to Python's original graph regardless of where it actually
     * sits on the grid.
     *
     * `exits` intentionally point INTO 14 ids this function itself
     * creates -- a fully closed loop, matching the Python source's
     * original topology exactly -- except Oakhaven's `west`, which is
     * deliberately left pointing at whatever real generated tile already
     * existed there before the graft (untouched by this function, since
     * it's not one of the 15 offsets below): the one gateway from the
     * tested home region out into Kotlin's much larger procedural world.
     */
    fun graft(locations: MutableMap<String, GameLocation>, anchorX: Int, anchorY: Int) {
        fun id(dx: Int, dy: Int) = "${anchorX + dx}_${anchorY + dy}"

        val idOakhaven = id(0, 0)
        val idWhisperingWoods = id(0, -1)
        val idDeepWoods = id(0, -2)
        val idShadowCaves = id(1, -1)
        val idSunkenCitadel = id(2, -1)
        val idOldRoad = id(1, 0)
        val idIronstoneFoothills = id(2, 0)
        val idIronstonePass = id(2, 1)
        val idDragonsPeak = id(2, 2)
        val idEasternSwamps = id(0, 1)
        val idCoastalTown = id(1, 1)
        val idSunkenShipwreck = id(1, 2)
        val idMountainPassNorth = id(3, 0)
        val idAncientRuins = id(-1, -2)
        val idForgottenCrypt = id(-1, -3)

        fun loc(
            locId: String, dx: Int, dy: Int, name: String, description: String,
            exits: Map<String, String>, populationTier: PopulationTier, difficultyTier: Int,
            beings: List<SpawnEntry> = emptyList(), items: List<Item> = emptyList(),
        ) {
            locations[locId] = GameLocation(
                id = locId, x = anchorX + dx, y = anchorY + dy, biome = Biome.PLAINS,
                name = name, description = description, populationTier = populationTier,
                difficultyTier = difficultyTier, difficultyScore = difficultyTier * 20,
                beings = beings, exits = exits, terrain = TerrainKind.LAND, items = items,
            )
        }

        val existingWestNeighbor = locations[id(-1, 0)]?.id ?: idOakhaven // falls back to a self-loop only in the pathological case that tile doesn't exist

        loc(
            idOakhaven, 0, 0, "Oakhaven Village",
            "A serene village surrounding a central square. The air smells of fresh bread and damp earth.",
            mapOf("north" to idWhisperingWoods, "east" to idOldRoad, "south" to idEasternSwamps, "west" to existingWestNeighbor),
            PopulationTier.CITY, difficultyTier = 1,
            beings = listOf(npc(ELDER_THERON, seed = 101), npc(OAKHAVEN_MERCHANT, seed = 102)),
        )
        loc(
            idWhisperingWoods, 0, -1, "Whispering Woods",
            "Dense forest with howling wind through misty trees. Ancient, gnarled trees loom overhead.",
            mapOf("south" to idOakhaven, "north" to idDeepWoods, "east" to idShadowCaves),
            PopulationTier.WILDERNESS, difficultyTier = 1,
            beings = listOf(direWolf(201), goblin(202)),
        )
        loc(
            idDeepWoods, 0, -2, "Deep Whispering Woods",
            "Pitch dark canopy where sunlight cannot reach. Strange sounds echo from the shadows.",
            mapOf("south" to idWhisperingWoods, "west" to idAncientRuins),
            PopulationTier.WILDERNESS, difficultyTier = 2,
            beings = listOf(goblin(203), direWolf(204)),
            items = listOf(questItem(ITEM_ANCIENT_RELIC, tier = 2, seed = 301)),
        )
        loc(
            idShadowCaves, 1, -1, "Shadow Caves",
            "Damp limestone cave network dripping with glowing flora. The air is cool and still.",
            mapOf("west" to idWhisperingWoods, "east" to idSunkenCitadel),
            PopulationTier.WILDERNESS, difficultyTier = 2,
            beings = listOf(goblin(205), giantToad(206)),
            items = listOf(material(ITEM_GLOWING_MUSHROOM, seed = 302), material(ITEM_GLOWING_MUSHROOM, seed = 303)),
        )
        loc(
            idSunkenCitadel, 2, -1, "Sunken Citadel Courtyard",
            "Ancient flooded stone fortress radiating dark power. Water laps at crumbling walls.",
            mapOf("west" to idShadowCaves),
            PopulationTier.CITY, difficultyTier = 3,
            beings = listOf(npc(ARCANE_VENDOR, seed = 103), shadowCultist(207), skeletalWarrior(208)),
        )
        loc(
            idOldRoad, 1, 0, "Old Forest Road",
            "Long dirt path connecting distant provinces. Wagon tracks are faintly visible.",
            mapOf("west" to idOakhaven, "east" to idIronstoneFoothills, "south" to idCoastalTown),
            PopulationTier.WILDERNESS, difficultyTier = 1,
            beings = listOf(goblin(209), banditOutlaw(210)),
        )
        loc(
            idIronstoneFoothills, 2, 0, "Ironstone Foothills",
            "Rocky ascending terrain surrounded by jagged cliff faces. The wind picks up here.",
            mapOf("west" to idOldRoad, "north" to idIronstonePass, "east" to idMountainPassNorth),
            PopulationTier.CITY, difficultyTier = 2,
            beings = listOf(npc(MOUNTAIN_GUIDE, seed = 104), npc(MOUNTAIN_KEEPER, seed = 105), banditOutlaw(211), mountainGoat(212)),
        )
        loc(
            idIronstonePass, 2, 1, "Ironstone Mountain Pass",
            "Freezing mountain path with howling blizzards. The air is thin and cold.",
            mapOf("south" to idIronstoneFoothills, "up" to idDragonsPeak),
            PopulationTier.WILDERNESS, difficultyTier = 4,
            beings = listOf(stoneGolem(213), mountainGoat(214)),
        )
        loc(
            idDragonsPeak, 2, 2, "Dragon's Peak Summit",
            "High volcanic summit covered in ancient scorch marks. A faint smell of sulfur lingers.",
            mapOf("down" to idIronstonePass),
            PopulationTier.WILDERNESS, difficultyTier = 5,
            beings = listOf(ancientDragon(215)),
        )
        loc(
            idEasternSwamps, 0, 1, "Eastern Swamps",
            "A murky, humid marshland. Strange plants and buzzing insects fill the air.",
            mapOf("north" to idOakhaven),
            PopulationTier.WILDERNESS, difficultyTier = 1,
            beings = listOf(giantToad(216), direWolf(217)),
            items = List(SWAMP_HERB_REQUIRED) { material(ITEM_SWAMP_HERB, seed = 310L + it) },
        )
        loc(
            idCoastalTown, 1, 1, "Coastal Town of Port Eldoria",
            "A bustling port town with the scent of salt and fish. Ships dock constantly.",
            mapOf("north" to idOldRoad, "east" to idSunkenShipwreck),
            PopulationTier.CITY, difficultyTier = 2,
            beings = listOf(npc(FISHERMAN_FINN, seed = 106), npc(COASTAL_TRADER, seed = 107)),
        )
        loc(
            idSunkenShipwreck, 1, 2, "Sunken Shipwreck",
            "The remains of a grand ship, half-submerged in shallow waters.",
            mapOf("west" to idCoastalTown),
            PopulationTier.WILDERNESS, difficultyTier = 2,
            beings = listOf(giantCrab(218)),
            items = listOf(questItem(ITEM_SHIP_MANIFEST, tier = 2, seed = 304)),
        )
        loc(
            idMountainPassNorth, 3, 0, "Northern Mountain Pass",
            "A narrow, icy path winding through towering peaks.",
            mapOf("west" to idIronstoneFoothills, "north" to idForgottenCrypt),
            PopulationTier.WILDERNESS, difficultyTier = 3,
            beings = listOf(mountainGoat(219), stoneGolem(220), npc(LOST_MINER, seed = 108)),
        )
        loc(
            idAncientRuins, -1, -2, "Ancient Ruins",
            "Crumbling stone structures overgrown with vines.",
            mapOf("east" to idDeepWoods, "north" to idForgottenCrypt),
            PopulationTier.WILDERNESS, difficultyTier = 3,
            beings = listOf(skeletalWarrior(221), shadowCultist(222), npc(ANCIENT_SCHOLAR, seed = 109)),
        )
        loc(
            idForgottenCrypt, -1, -3, "Forgotten Crypt",
            "A dark, musty crypt beneath the ancient ruins.",
            mapOf("south" to idAncientRuins, "west" to idMountainPassNorth),
            PopulationTier.WILDERNESS, difficultyTier = 4,
            beings = listOf(skeletalWarrior(223)),
            items = listOf(questItem(ITEM_ANCIENT_COMPASS, tier = 3, seed = 305)),
        )
    }
}
